modules/gapi/src/backends/common/serialization.cpp

```diff
@@ -447,12 +505,49 @@ I::IStream& operator>> (I::IStream& is, cv::gimpl::Op &op) {
 I::OStream& operator<< (I::OStream& os, const cv::gimpl::Data &d) {
     // FIXME: HostCtor is not stored here!!
     // FIXME: Storage may be incorrect for subgraph-to-graph process
-    return os << d.shape << d.rc << d.meta << d.storage;
+    return os << d.shape << d.rc << d.meta << d.storage << d.kind;
 }
+
+namespace
+{
+    template<typename Ref, typename T, typename... Ts>
+    struct initCtor;
+
+    // FIXME: workaround to end the recursion
+    template<typename Ref>
+    struct initCtor<Ref, std::tuple<>>
+    {
+        static void init(cv::gimpl::Data&)
+        {}
+    };
+
+    template<typename Ref, typename T, typename... Ts>
+    struct initCtor<Ref, std::tuple<T, Ts...>>
+    {
+        static void init(cv::gimpl::Data& d) {
+            if (d.kind == cv::detail::GOpaqueTraits<T>::kind) {
+                static std::function<void(Ref&)> ctor = [](Ref& ref){ref.template reset<T>();};
+                d.ctor = ctor;
+            } else {
+                initCtor<Ref, std::tuple<Ts...> >::init(d);
+            }
+        }
+    };
+} // anonymous namespace
+
 I::IStream& operator>> (I::IStream& is, cv::gimpl::Data &d) {
     // FIXME: HostCtor is not stored here!!
     // FIXME: Storage may be incorrect for subgraph-to-graph process
-    return is >> d.shape >> d.rc >> d.meta >> d.storage;
+    is >> d.shape >> d.rc >> d.meta >> d.storage >> d.kind;
+    if (d.shape == cv::GShape::GARRAY)
+    {
+        initCtor<cv::detail::VectorRef, cv::detail::GOpaqueTraitsArrayTypes>::init(d);
+    }
+    else if (d.shape == cv::GShape::GOPAQUE)
+    {
+        initCtor<cv::detail::OpaqueRef, cv::detail::GOpaqueTraitsOpaqueTypes>::init(d);
+    }
```

Comment:
else assert?

Reply:
No, we only need to initialize ctor for array and opaque here. For other types logic remains the same